File: frontend/src/types/dataset.ts
export type DatasetTaskType = "classification" | "regression" | "multi_output_classification";

export type Dataset = {
  id: number;
  user_id: number;
  name: string;
  task_type: DatasetTaskType;
  description: string | null;
  file_path: string | null;
  file_type: string | null;
  sample_count: number;
  feature_count: number;
  target_columns: string[];
  created_at: string;
  updated_at: string;
};

export type DatasetColumn = {
  id: number;
  dataset_id: number;
  column_name: string;
  data_type: string;
  role: string;
  missing_count: number;
  unique_count: number;
  mean: number | null;
  std: number | null;
  min_value: number | null;
  max_value: number | null;
};

export type DatasetCreatePayload = {
  name: string;
  task_type: DatasetTaskType;
  description?: string;
};

export type DatasetPreview = {
  columns: string[];
  rows: Record<string, unknown>[];
  total_rows: number;
};

export type DatasetProfile = {
  dataset: Dataset;
  columns: DatasetColumn[];
  missing_values: Record<string, number>;
  target_distribution: Record<string, Record<string, number>>;
};

File: backend/app/db_models/dataset.py
from datetime import datetime
from typing import Any

from sqlalchemy import DateTime, Float, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Dataset(Base):
    __tablename__ = "datasets"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True, nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    task_type: Mapped[str] = mapped_column(String(64), default="classification", nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    file_path: Mapped[str | None] = mapped_column(String(512), nullable=True)
    file_type: Mapped[str | None] = mapped_column(String(16), nullable=True)
    sample_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    feature_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    target_columns: Mapped[list[str]] = mapped_column(JSON, default=list, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
    )


class DatasetColumn(Base):
    __tablename__ = "dataset_columns"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    dataset_id: Mapped[int] = mapped_column(ForeignKey("datasets.id"), index=True, nullable=False)
    column_name: Mapped[str] = mapped_column(String(255), nullable=False)
    data_type: Mapped[str] = mapped_column(String(64), nullable=False)
    role: Mapped[str] = mapped_column(String(32), default="feature", nullable=False)
    missing_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    unique_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    mean: Mapped[float | None] = mapped_column(Float, nullable=True)
    std: Mapped[float | None] = mapped_column(Float, nullable=True)
    min_value: Mapped[float | None] = mapped_column(Float, nullable=True)
    max_value: Mapped[float | None] = mapped_column(Float, nullable=True)

    def to_summary(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "dataset_id": self.dataset_id,
            "column_name": self.column_name,
            "data_type": self.data_type,
            "role": self.role,
            "missing_count": self.missing_count,
            "unique_count": self.unique_count,
            "mean": self.mean,
            "std": self.std,
            "min_value": self.min_value,
            "max_value": self.max_value,
        }
